### PR TITLE
Update aten_fill to avoid d2h sync

### LINKED ISSUE #109115: `aten::fill_` causes H2D sync
`aten::fill_` calls into `aten::item`, which causes a H2D sync. Can we implement `fill_` without this sync?

Example repro:
```
a = torch.tensor(0, device="cuda")
x = torch.rand((100,), device="cuda")
amax = torch.max(torch.abs(x))
a.fill_(amax)
```
`amax` is a 0-dim tensor (i.e. with size `torch.Size([])`) as required by `fill_`.

Screenshot 2023-09-12 at 12 22 31 PM

Workaround to avoid the H2D sync:
```
a *= 0
a += amax
```

cc @ptrblck

### PR DESCRIPTION
**Pull request by drisspg**
Fixes #109115

### Before:
Screenshot 2023-09-18 at 11 57 32 AM

### After:
Screenshot 2023-09-18 at 11 57 25 AM

cc @ptrblck

### CODE REVIEW

**Comment by drisspg** [comment]
I am also losing my mind:
Testing Script
```Python
import torch

from transformer_nuggets.utils.benchmark import ProfileConfig, profile_function

def main():
    a = torch.tensor(0, device="cuda")
    x = torch.rand((100,), device="cuda")
    amax = torch.max(torch.abs(x))
    # profile_config = ProfileConfig("/home/drisspg/meta/scripts/fp8/data/aten_fill.json", name="aten_fill_scaler", iters=3, sync=True)
    # profile_function(profile_config,torch.fill_, a, amax)
    print(a)
    print(amax)
    a.fill_(amax)
    print(a)
    print(amax)
    torch.fill_(a, amax)
    print(a)
    print(amax)

if __name__ == "__main__":
    main()
```
``` Shell
❯ python /home/drisspg/meta/scripts/fp8/aten_fill.py
tensor(0, device='cuda:0')
tensor(0.9860, device='cuda:0')
tensor(0, device='cuda:0')
tensor(0.9860, device='cuda:0')
tensor(0, device='cuda:0')
tensor(0.9860, device='cuda:0')
```

File: aten/src/ATen/native/Fill.cpp [hunk 1]
```diff
@@ -40,7 +40,7 @@ Tensor& fill_out(Tensor& self, const Scalar& value) {
 static Tensor& fill_out_quantized(Tensor& self, const Scalar& value) {
   at::Tensor out = at::ones(self.sizes()).to(kFloat) * value;
```
Comment (vadimkantorov) [inline aten/src/ATen/native/Fill.cpp:41]: btw this looks like allocating quite a lot :) given that `self` will get filled, and `out` thrown away

Comment (drisspg) [reply aten/src/ATen/native/Fill.cpp:41]: agreed but I don't really know where out_quantized is being sued from and the reasons why it was written in this way

Comment (drisspg) [reply aten/src/ATen/native/Fill.cpp:41]: that being said: https://github.com/pytorch/pytorch/pull/109544

**Comment by vadimkantorov** [comment]
Also, `fill_scaler` is a typo for `fill_scalar`?

File: test/nn/test_embedding.py [hunk 2]
```diff
@@ -602,7 +602,8 @@ def embedding_bag_check(indices, weights, mode, sparse, padding_idx):
                     # Fill one row with duplicate index so we can test with a fully
                     # padded row
                     duplicate_row = random.randint(0, indices_dim0 - 1)
-                    indices[duplicate_row] = indices[duplicate_row][0]
+                    fill_value = indices[duplicate_row][0].clone()
```
Comment (mikaylagawarecki) [inline test/nn/test_embedding.py:605]: Commenting for posterity that this change is because `copy_` uses TensorIterator which does [`assert_no_partial_overlap`](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/MemoryOverlap.cpp#L76-L78) and `fill_` is called in `copy_to` [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/TensorIndexing.h#L409-L411)

**Review by mikaylagawarecki — APPROVED** [review]

**Comment by vadimkantorov** [comment]
> Also, `aten_fill_scaler` is a typo for `aten_fill_scalar`? (looking at your profiler screenshots)

@drisspg @mikaylagawarecki hmm could not find `aten_fill_scaler` in the source code

File: aten/src/ATen/native/Fill.cpp [hunk 3]
```diff
@@ -55,7 +55,8 @@ Tensor& fill_quantized_(Tensor& self, const Scalar& value) {
 
 Tensor& fill_(Tensor& self, const Tensor& value) {
   TORCH_CHECK(value.dim() == 0, "fill_ only supports 0-dimension value tensor but got tensor with ", value.dim(), " dimensions.");
-  return fill_out(self, value.item());
+  self.copy_(value);
```
Comment (ngimel) [inline aten/src/ATen/native/Fill.cpp:58]: mm, no? copy_ is synchronizing most of the time (when `value` is on the host and `self` on the device)

File: test/nn/test_embedding.py (hunk 2 above)
Comment (ngimel) [reply test/nn/test_embedding.py:605]: So this was a bc-breaking change?

File: aten/src/ATen/native/Fill.cpp (hunk 3 above)
Comment (drisspg) [reply aten/src/ATen/native/Fill.cpp:58]: so should I update this to only copy when value is on device? otherwise call fill_out?

Comment (ngimel) [reply aten/src/ATen/native/Fill.cpp:58]: Yes, that'll give you best of both worlds

File: test/nn/test_embedding.py (hunk 2 above)
Comment (mikaylagawarecki) [reply test/nn/test_embedding.py:605]: Yea we missed this, Driss later reverted this in https://github.com/pytorch/pytorch/pull/109835
